### ISSUE TITLE
JUnit parallel executor running too many tests with a fixed policy.

### ISSUE THREAD

**Issue by OPeyrusse** [junit-team/junit5#3108]
JUnit only parallel executor service relies on a ForkJoinPool. Unfortunately, this does not play well with code also using ForkJoinPools.

The observed issue is that, when activating parallel tests, JUnit uses `ForkJoinPoolHierarchicalTestExecutorService`. However, our tests are also using `ForkJoinPool`s and `ForkJoinTask`s. The orchestration of the test awaits for the completion of those tasks before moving on in the tests.
But the issue is that `ForkJoinTask` and `ForkJoinWorkerThread` are capable of detecting the use of the FJP framework ([here](https://github.com/openjdk/jdk/blob/98fa48c330941efe6588a907b383802a11ed0e6b/src/java.base/share/classes/java/util/concurrent/ForkJoinTask.java#L411-L415) for example) and react to it. As JUnit tasks and the project tasks are in different ForkJoinPools, they cannot help each other. This only results in more tests being started by already running and **incomplete** tests.

This can be an issue when tests are resource sensitive. For example, it may not be possible to open too many connections to a Database.
Tough not explicitly illustrated in this project, we also faced `StackOverflowError` because of recursive test executions in a single worker. In the following logs, produced by the reproducing project, we can see that two workers are recursively starting tests before finishing any:
```console
ForkJoinPool-1-worker-1 starting a new test. Running: 2
ForkJoinPool-1-worker-2 starting a new test. Running: 1
ForkJoinPool-1-worker-1 starting a new test. Running: 3
ForkJoinPool-1-worker-1 starting a new test. Running: 4
ForkJoinPool-1-worker-1 starting a new test. Running: 5
ForkJoinPool-1-worker-2 starting a new test. Running: 6
ForkJoinPool-1-worker-1 starting a new test. Running: 7
ForkJoinPool-1-worker-1 starting a new test. Running: 8
ForkJoinPool-1-worker-1 starting a new test. Running: 9
ForkJoinPool-1-worker-2 starting a new test. Running: 10
ForkJoinPool-1-worker-1 starting a new test. Running: 11
ForkJoinPool-1-worker-1 starting a new test. Running: 12
ForkJoinPool-1-worker-1 starting a new test. Running: 14
ForkJoinPool-1-worker-2 starting a new test. Running: 13
ForkJoinPool-1-worker-1 starting a new test. Running: 15
ForkJoinPool-1-worker-1 starting a new test. Running: 16
ForkJoinPool-1-worker-1 starting a new test. Running: 17
ForkJoinPool-1-worker-1 starting a new test. Running: 18
ForkJoinPool-1-worker-2 starting a new test. Running: 19
ForkJoinPool-1-worker-1 starting a new test. Running: 20
```
> worker-1 started 15 tests, worker-2 started 5 tests

In actual code, given the location of the point of respawn, this can generate very large stacks.

An alternative implementation of the executor service, as shown in [this PR](https://github.com/junit-team/junit5/pull/2805), using a standard Thread Executor, would not show similar issues, at the expense of not ideally orchestrating multiple executions.

Let's note that this is a very sneaky issue. Even in this project, we can see that the call to `ForkJoinTask#get` is not visible in the JUnit method. And it can be as deep as we want in the stack, even hidden to some users as it is happening in a used framework or tool. It may not be always possible for a user to detect that pattern.

## Steps to reproduce

See this project https://github.com/OPeyrusse/junitforkjoinpool
After building the project, run the test class `TestCalculator` (or look at the README if changed since opening this issue).

## Context

 - Used versions (Jupiter/Vintage/Platform): 5.8.2, with the Jupiter runner
 - Build Tool/IDE: Build with Intellij, failing in Intellij and when running it with Maven
 - Maven Surefire version: 3.0.0-M5

## Deliverables

- [ ] At least one standard test executor not relying on the ForkJoinPool

**Comment by OPeyrusse** [comment]
Some additional thoughts: 
Looking at the code and being used to work with ForkJoinPool, I can guess some benefits of using the ForkJoin framework. It would naturally focus on sub-tasks to complete the execution of a given class or parameterized tests before moving to other classes. 
However, IMO, the mistake is that a `ForkJoinTask` should only run CPU intensive tasks. JUnit cannot know the content of tests. Some may be real unit tests, only depending on the CPU. Some may be real unit tests but reading the disk. Some may be more integration tests, making network calls using whatever framework. In those cases, for defensive measures, users of the ForkJoin framework have to rely on [ManagedBlock](https://github.com/openjdk/jdk/blob/98fa48c330941efe6588a907b383802a11ed0e6b/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L3718-L3726). The drawback of Managed blocker is that they start new threads.

Yet another solution could be to use a ForkJoinPool for scheduling of the tests and a standard executor for the execution of the tests.

**Comment by OPeyrusse** [comment]
There is also a short snippet reproducing this behavior [here](https://github.com/junit-team/junit5/pull/3044#issuecomment-1290523279).
However, the current choice was to say that the doc will be update to say the configuration limits the number of threads, not the number of tests. This issue focuses on the number of tests being run, as it is the actual problem at stake, leading to `StackOverflowError`s.

**Comment by OPeyrusse** [comment]
To add to this, it is not currently possible to have parallel tests without being tricked by JUnit current implementation.
It is either sequential tests or parallel tests that must take care of what they do.

**Comment by marcphilipp** [comment]
@OPeyrusse Do the [features released as part of 5.9.2](https://junit.org/junit5/docs/5.9.2/release-notes/index.html#new-features-and-improvements-2) help in your case?

**Comment by OPeyrusse** [comment]
Hello @marcphilipp , nope it will not help. The feature advertised was already one we looked at, when it was at the stage of PR.
If you look again, you will see that only two threads from the ForkJoinPool are being used.
Tough not tested, I would even say that with only one thread, you can reproduce the problem above. As stated in my description, it comes from the FJ framework detecting the use of ForkJoinPool in a ForkJoinThread, leading it to attempt some work instead of waiting.
So limiting the number of threads in the pool or its saturation will not resolve the issue. Here, the issue is that too many tests are started.

I can give you an analogy with the following story:
There is a factory that can manufacture cars. It takes time, more or less depending on the car models. Building a car is the analogy to running a single test. The factory has enough workers to build several cars concurrently.
There are also clerks coming to the factory with orders to manufacture cars. Those are JUnit ForkJoinTasks, produced after discovering that tests must be run.
When a clerk arrives at the factory, it delivers the order to the factory, that immediately starts working on it. 
The factory has chosen, for company reasons, to operate using the FJP methodology (they are smart). The weird thing is that the clerk **sees** that, and because the clerk company is also using the FJP methodology, the clerk decides not to stay idle and do something while waiting for the car to be made. Obviously, the clerk cannot join the factory workers - it is not its job - so to do something, the clerk picks the next task available to him, which is **go back to the office to pick another order**. 
Because clerks make faster trips to the office than the factory can produce cars, the factory becomes overflowed with orders - which, in the computer world, translates to memory starvation, stack overflow, etc

I hope this helps understand the logic triggering the issue :smiley:

**Comment by marcphilipp** [comment]
Nice analogy. I wish the car factory wouldn't leak implementation details... 😉 

> Yet another solution could be to use a ForkJoinPool for scheduling of the tests and a standard executor for the execution of the tests.

I think we can consider that. Would you like to give it a shot?

**Comment by marcphilipp** [comment]
@OPeyrusse In `ParallelExecutionIntegrationTests` we have tests that check that tests are run on different threads by collecting the names of those threads (see the nested `ThreadReporter` class). I think for this, all we'd need to verify is that they are not `ForkJoinWorkerThread`.

Generally, I think we should make this new behavior opt-in via a configuration parameter since it introduces additional overhead.

**Comment by OPeyrusse** [comment]
Thanks for the tips. I will look at it in the coming days.

**Comment by OPeyrusse** [comment]
Just a quick update: I have not abandoned nor forgotten this issue. It is simply that I am currently taking care of my children during the holiday season here. I already have ideas on how to proceed and will start on this next week :smile:

**Comment by asardaes** [comment]
If I may ask, do you think it would be possible to allow custom `HierarchicalTestExecutorService` implementations? I imagine it would be possible to simply extend the configuration of the `dynamic` strategy to allow choosing between predefined implementations, but I would personally like more control over the thread pool in a programmatic way. I think the `custom` strategy could then check if `junit.jupiter.execution.parallel.config.custom.class` is an instance of `ParallelExecutionConfigurationStrategy` or `HierarchicalTestExecutorService`, or something like that.

**Comment by marcphilipp** [comment]
@asardaes That's an interesting idea. Could you please raise a new issue for it?

**Comment by asardaes** [comment]
@marcphilipp since you mentioned @OPeyrusse's implementation should be opt-in, I think it's worth discussing in the context of this ticket, killing 2 birds with 1 stone. I gave it some more thought, let me explain what I have in mind.

A new configuration parameter could be added, say `junit.jupiter.execution.parallel.executor.class`. If this is not set, it defaults to either `org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService` or  `org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService` depending on the value of `junit.jupiter.execution.parallel.enabled`. With this new parameter, users could either choose the "standard executor" implemented by @OPeyrusse or a custom one they themselves implement; if the users specify a custom one, they could decide whether to ignore  `junit.jupiter.execution.parallel.enabled` or not.

To instantiate `junit.jupiter.execution.parallel.executor.class`, all implementations would need some convention that could be used by `JupiterTestEngine` (and any other custom engine that wants to support different executors). I imagine it would be easiest to require them to have a no-arg constructor, and then adding a default method `configure(ConfigurationParameters)` method to `HierarchicalTestExecutorService`. This would allow custom implementations to read custom configuration parameters, but it would mean the executors included out of the box in JUnit would have to check if `junit.jupiter.execution.parallel.config.strategy=custom` and instantiate `junit.jupiter.execution.parallel.config.custom.class` themselves.

I'm not so sure about the instantiation/configuration convention. Given that both included executor services are marked as stable, I imagine they need to keep existing constructors, but adding a new one and having a default `configure` method in the interface should be fine, no?

**Comment by OPeyrusse** [comment]
@asardaes do you have a special use-case in mind, driving your question?

I am asking because my idea for this issue was to have two different ways of running the tests: one inlining the execution withing the ForkJoinPool, and one submitting a task to an external Executor and waiting for the result.
Behind this, though I don't know the historical reasons for creating a test executor on top of the ForkJoinPool, I assume that it comes from its work-stealing feature. It can benefit to test execution when you launch a group of tests - often those under a class - as the executor should focus on executing those tests in parallel instead of starting the tests of the next class.
I want to keep this hypothetical design choice. Therefore, my task is only to separate the way one individual test is running.
If we wanted some extension, I could open this to load a user-selected class, from its name, as you suggested. But it would only affect individual test execution, not the general scheduling of all the tests.
Would that suits your needs?

**Comment by asardaes** [comment]
@OPeyrusse no, I don't think that would fit what I had in mind, but if your implementation doesn't need a completely new `HierarchicalTestExecutorService` but rather does something else inside an existing one, then I guess my idea really would require a separate issue.

Peesonally I'm just experimenting with kotlin coroutines, so I don't need a custom test engine, I just want a custom executor.

**Comment by OPeyrusse** [comment]
@asardaes If you say so. Not being a core developer of JUnit, I cannot decide what to do. I will let @marcphilipp decide what to do and will start going forward with my plan, that is not creating a new executor.

**Comment by marcphilipp** [comment]
@asardaes I think what you describe in https://github.com/junit-team/junit5/issues/3108#issuecomment-1537380013 makes a lot of sense. However, as you said in https://github.com/junit-team/junit5/issues/3108#issuecomment-1540593193, this issue seems more like making an existing executor behave differently based on a different configuration parameter. So I think a new issue for your use case would be the way to go.

**Comment by OPeyrusse** [comment]
It is now obvious to me that I will not find time to work on this in the coming special. My personal situation is special but has a stable rhythm, and that rhythm does not allow me enough time to work on this issue.
While I am still very interested, I prefer play the card of honesty and say that I cannot commit on work for now.
When there is a change in my schedule, I will come back and see if help is still needed :smile:

**Comment by marcphilipp** [comment]
@OPeyrusse No worries and thanks for letting us know!

Is anyone else interested in picking it up in the meantime?

**Comment by shans96** [comment]
Hi, keen to try resolving this issue. Am I right in saying that this functionality should be configured via `ConsoleLauncher`?

**Comment by marcphilipp** [comment]
@shans96 It should be opt-in via a new configuration parameter so it would work from the `ConsoleLauncher` but also other use cases such as IDEs and build tools.

**Comment by Vampire** [comment]
Please also have in mind, that this is not something Jupiter specific.
I right now have a similar / related problem.
I'm using Spock (another JUnit platform engine) to run some tests.
These tests (or rather one component of the SUT) heavily use `CompletableFuture`s and `join()` on them to then verify the result.
This means while the `parallelism` is set to `6`, basically all tests are run in parallel as all tests quickly get to a `join()` which causes further tests to be started.
This frys the CPU and makes the whole computer unusable.
For some reason setting the `maxPoolSize` does not really help. It caused less tests to run in paralle, but it also seems to just lead to some deadlock or at least I was too impatient to wait for the test to continue.

With the solution suggested here above, running the actual test on a different thread pool, it works much better.
I simulated it using a Spock iteration interceptor that does the actual test execution on a different thread and waits for it to finish without the FJP starting additional threads.
With that it works as expected.

From a user perspective, this default behavior was highly confusing and non-obvious, and it will hit anyone who is testing something using `ForkJoinPool.managedBlock` transitively like `CompletableFuture#join` or `CompletableFuture#get` or `Phaser` or whatever.
So maybe it could be reconsidered whether the new behavior should become the default, at least if the additional overhead is not too significant.

**Comment by asardaes** [comment]
@OPeyrusse @shans96 are you certain [the changes mentioned above](https://junit.org/junit5/docs/5.9.2/release-notes/index.html#new-features-and-improvements-2) won't help? (They will soon also apply to the dynamic strategy, see #3206) I did some experiments and it seems to me that if you set the maximum pool size equal to parallelism and allow saturation, the fork-join pool will not compensate. I'm not sure right now if it matters, but I set `minRunnable=0` when instantiating the pool. This of course assumes Java 9+.

**Comment by OPeyrusse** [comment]
Hello @asardaes, yes I am absolutely sure that the solution of #3206 does not help because there is no compensation mechanism nor creation of new thread nor anything like that at play.
If you run [my reproducing project](https://github.com/OPeyrusse/junitforkjoinpool) and look at the stacktrace, or just look again at the output in the OP, you will see that a single thread is starting many tests without completing anyone.
That's because new tests are started when calling `ForkJoinTask#get` or `ForkJoinTask#join` **inside a test**. The **same** thread is beginning many tests. So no, limiting the number of threads will not solve this problem.

It is also worth noting that adding a mechanism that would limit the number of tests running simultaneously - using something like a `Semaphore` - **would not work**. It would create a deadlock because the thread would be block at some point and would have to unroll the stack to continue working on the tests already started.
_(A drawing would really help here but I don't have right now the tools to do it. I will try later)_
To me, this is a limitation of the ForkJoin framework, not a flaw. One must have it in mind when using this framewok.

So, again, to answer your post, no, it won't help

**Comment by Vampire** [comment]
I can also confirm again like I just did in my last comment. Setting the max pool size and saturation does not work. The only way I could mitigate was to make sure the actual test execution is run in a separate thread pool that is not a `ForkJoinPool` and that has the max threads set to parallelism, so that the test scheduling can use the work-stealing, but the test execution is not.

In my Spock-using project I have a Spock extension that creates a thread pool using `threadPool = Executors.newFixedThreadPool(runnerConfiguration.parallel.parallelExecutionConfiguration.parallelism)` and adds an iteration interceptor that does
```groovy
spec.allFeatures*.addIterationInterceptor {
    threadPool.submit(it::proceed).get()
}
```
and with that it behaves as expected, effectively doing the solution suggested here above.

You can most probably do something similar for Jupiter tests.
Looking through the Jupiter docs I found https://junit.org/junit5/docs/current/user-guide/#extensions-intercepting-invocations where it is demonstrated how to do the actual test execution in the Swing EDT. This is exactly what you also need here for the work-around I used in Spock, so probably something like
```java
@Override
public void interceptTestMethod(Invocation<Void> invocation,
        ReflectiveInvocationContext<Method> invocationContext,
        ExtensionContext extensionContext) throws Throwable {
    threadPool.submit(invocation::proceed).get();
}
```

**Comment by Vampire** [comment]
Just to clarify again.
In my tests no other `ForkJoinPool` is involved.
Just `CompletableFuture`s that behave differently if waited for on a fork-join thread, and also the locks for exclusive resources.

**Comment by asardaes** [comment]
To clarify, I think it's fine to implement a different approach if possible (in fact, I've experimented a bit with that as well), I just thought it would be good to avoid it if possible. I've looked through all the comments and mentioned code, but maybe there's more to the other arguments in the `ForkJoinPool` constructor. Here's an experiment I did (in Kotlin, I hope that's ok):

```kotlin
class SleepingAction(private val sleepMs: Long) : RecursiveAction() {
    override fun compute() {
        println("${Instant.now()} - sleeping $sleepMs ms in thread ${Thread.currentThread().id}")
        Thread.sleep(sleepMs)
        println("${Instant.now()} - woke up after $sleepMs ms in thread ${Thread.currentThread().id}")
    }
}

class SecondPoolAction(private val secondPool: ForkJoinPool) : RecursiveAction() {
    override fun compute() {
        secondPool.submit(object : RecursiveAction() {
            override fun compute() {
                try {
                    val sleepingActions = listOf(
                        SleepingAction(1_000L),
                        SleepingAction(2_000L),
                    )
                    println("${Instant.now()} - nested forking in thread ${Thread.currentThread().id}")
                    sleepingActions.forEach { it.fork() }
                    sleepingActions.reversed().forEach { it.join() }
                    println("${Instant.now()} - nested done in thread ${Thread.currentThread().id}")
                } finally {
                    secondPool.shutdownNow()
                }
            }
        }).get()
    }
}

fun main() {
    val pool = ForkJoinPool(2, ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, false, 0, 3, 0, { true }, 10L, TimeUnit.SECONDS)
    val secondPool = ForkJoinPool(2, ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, false, 0, 2, 0, { true }, 10L, TimeUnit.SECONDS)
    try {
        pool.invoke(object : RecursiveAction() {
            override fun compute() {
                val sleepingActions = listOf(
                    SecondPoolAction(secondPool),
                    SleepingAction(1_500L),
                    SleepingAction(4_000L),
                )
                println("${Instant.now()} - forking in thread ${Thread.currentThread().id}")
                sleepingActions.forEach { it.fork() }
                sleepingActions.reversed().forEach { it.join() }
                println("${Instant.now()} - done in thread ${Thread.currentThread().id}")
            }
        })
    } finally {
        pool.shutdownNow()
    }
}
```

And one potential outcome:

```
2023-05-25T17:58:21.392305940Z - forking in thread 16
2023-05-25T17:58:21.402393533Z - nested forking in thread 18
2023-05-25T17:58:21.403428288Z - sleeping 1000 ms in thread 19
2023-05-25T17:58:21.402992225Z - sleeping 4000 ms in thread 16
2023-05-25T17:58:21.403326525Z - sleeping 2000 ms in thread 18
2023-05-25T17:58:22.407123722Z - woke up after 1000 ms in thread 19
2023-05-25T17:58:23.407246362Z - woke up after 2000 ms in thread 18
2023-05-25T17:58:23.407451367Z - nested done in thread 18
2023-05-25T17:58:23.408582931Z - sleeping 1500 ms in thread 17 // <--------------------------
2023-05-25T17:58:24.908845221Z - woke up after 1500 ms in thread 17
2023-05-25T17:58:25.407028902Z - woke up after 4000 ms in thread 16
2023-05-25T17:58:25.407199927Z - done in thread 16
```

Note that the line I point to doesn't start immediately after the second pool's future's `get` is called.

From what I see, `maximumPoolSize` could even be larger, the critical thing seems to be `minimumRunnable` - setting it to 0 or 1 behaves as above, but anything larger than that causes the same problem discussed in this issue.

**Comment by Vampire** [comment]
I tried with removing my work-around and setting `minimumRunnable` to `0`.
It behaved "different", but still not as expected. The amount of tests increased slower, but still crept up to 16 or so with a parallelism of 6.
And at some point everything was hanging and nothing continued.

**Comment by asardaes** [comment]
I know @shans96 also started looking at a fix for this, but I figured I can tell all of you about the PoC I've been experimenting with; you can see the commit linked above.

I wrote a couple of TODOs there, but they're both related to my understanding of `CONCURRENT` execution mode semantics. The current implementation checks execution mode in both `submit` and `forkConcurrentTasks` + `executeNonConcurrentTasks`, but in both scenarios, if the task is *not* concurrent, it executes it in the current thread, and that doesn't know if some other thread in the pool has stolen work and is executing another test, so is it really guaranteeing that no other test is running at the same time?

I added a method `ExclusiveTask#await` and use it instead of `join` in `joinConcurrentTasksInReverseOrderToEnableWorkStealing` - if the task should execute in a `ForkJoinWorkerThread`, `await` delegates to `join`, otherwise it delegates to `compute`. If I don't do it like this, the tests execute fine and are even show as PASSED in the terminal, but after that everything stalls and all threads in the fork-join pool remain parked, and I cannot figure out why. I tried to reproduce this outside of a JUnit 5 session and I couldn't.

**Comment by Vampire** [comment]
Doesn't it simply work like suggested above:
> Yet another solution could be to use a ForkJoinPool for scheduling of the tests and a standard executor for the execution of the tests.

Which is exactly what I did in my described work-around, creating a fixed thread pool with `parallelism` size and doing the actual test execution on there? So far this works perfectly fine here.

**Comment by Vampire** [comment]
I'm pretty sure I'm just too naive. :-D
I hoped it would more or less be done with executing the `Node#around` action in said extra thread pool or something like that. :-D

**Comment by OPeyrusse** [comment]
> @OPeyrusse Hi, any updates for this issue? Looking forward for your good news. I've been struggling about three weeks for this issue...

Hello @mufasa1 , as stated in this [previous comment](https://github.com/junit-team/junit5/issues/3108#issuecomment-1536165905), I stopped working on this isssue and since then, @shans96 offered to take it and already tried some fixes. I will let them comment on their progress :smiley:

**Comment by mufasa1** [comment]
@OPeyrusse  Thanks for your reminder.  @shans96 Hi, is there any updates for this issue?

**Comment by shans96** [comment]
Hi, apologies for the delay in reply on this one. I have been looking into this but could use some guidance; as a starting point I've been trying to find ways of writing some unit tests that check  `ParallelExecutionIntegrationTests` to confirm that the new executor will work. However, I haven't been able to fully figure out how the integration tests work- i.e., how to call them with a specific executor. I think if I'm able to do that, it should be a lot easier to iterate on building a new test executor, so any advice would be welcome. 

That said, I think the commit that @asardaes linked is further ahead than what I've done so far. Not sure if they're still working on it?

**Comment by asardaes** [comment]
Mm I wouldn't mind submitting a PR myself, but I also need some guidance from the maintainers with the questions I mentioned above.

**Comment by mufasa1** [comment]
@shans96 
Hi, got this info from document and may I ask if it is possible to limt the maximum number of threads via jdk 8？or it can only be implemented in jdk9+.
![image](https://github.com/junit-team/junit5/assets/22904776/dd609434-730f-43fb-a409-3464c3422eb8)

**Comment by shans96** [comment]
Hi @mufasa1, I don't think the JDK 8 version of `ForkJoinPool`  supports limiting the maximum number of threads, in [JDK 9+](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/util/concurrent/ForkJoinPool.html#ForkJoinPool-int-java.util.concurrent.ForkJoinPool.ForkJoinWorkerThreadFactory-java.lang.Thread.UncaughtExceptionHandler-boolean-int-int-int-java.util.function.Predicate-long-java.util.concurrent.TimeUnit-) there's a constructor which takes a parameter `maximumPoolSize`, which does what you're looking for. `ForkJoinPool` in [JDK 8](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html) doesn't have the same one.

**Comment by igogoka** [comment]
> I can also confirm again like I just did in my last comment. Setting the max pool size and saturation does not work. The only way I could mitigate was to make sure the actual test execution is run in a separate thread pool that is not a `ForkJoinPool` and that has the max threads set to parallelism, so that the test scheduling can use the work-stealing, but the test execution is not.
> 
> In my Spock-using project I have a Spock extension that creates a thread pool using `threadPool = Executors.newFixedThreadPool(runnerConfiguration.parallel.parallelExecutionConfiguration.parallelism)` and adds an iteration interceptor that does
> 
> ```groovy
> spec.allFeatures*.addIterationInterceptor {
>     threadPool.submit(it::proceed).get()
> }
> ```
> 
> and with that it behaves as expected, effectively doing the solution suggested here above.
> 
> You can most probably do something similar for Jupiter tests. Looking through the Jupiter docs I found https://junit.org/junit5/docs/current/user-guide/#extensions-intercepting-invocations where it is demonstrated how to do the actual test execution in the Swing EDT. This is exactly what you also need here for the work-around I used in Spock, so probably something like
> 
> ```java
> @Override
> public void interceptTestMethod(Invocation<Void> invocation,
>         ReflectiveInvocationContext<Method> invocationContext,
>         ExtensionContext extensionContext) throws Throwable {
>     threadPool.submit(invocation::proceed).get();
> }
> ```
Please, write an example how I can set your sollution

**Comment by Vampire** [comment]
You just quoted the example?

**Comment by igogoka** [comment]
> You just quoted the example?
Where I should put this code "threadPool = Executors.newFixedThreadPool(runnerConfiguration.parallel.parallelExecutionConfiguration.parallelism)" to run fixed thread pools

**Comment by Vampire** [comment]
Wherever you need it.
Depends on how you adapt the work-around, for example which test engine you are using.
In my Spock project I have it in the `start()` method of a global extension.

**Comment by igogoka** [comment]
I use groovyVersion = '3.0.19', spockVersion = '2.4-M1-groovy-3.0',  gebVersion = '6.0', seleniumVersion = '4.17.0'. In SpockConfig.groovy :
`runner {
    parallel {
        enabled true
        fixed 4
    }
}`
But "fixed 4" doesn't work. So you override start() method from IGlobalExtension? Can you show how did you that with all imports, please.

**Comment by igogoka** [comment]
> Wherever you need it. Depends on how you adapt the work-around, for example which test engine you are using. In my Spock project I have it in the `start()` method of a global extension.

Can you show how you  Override start() method?

**Comment by Vampire** [comment]
```groovy
class FixJunit3108Extension implements IGlobalExtension {
    private static ExecutorService threadPool

    private final RunnerConfiguration runnerConfiguration

    FixJunit3108Extension(RunnerConfiguration runnerConfiguration) {
        this.runnerConfiguration = runnerConfiguration
    }

    @Override
    void start() {
        threadPool = Executors.newFixedThreadPool(runnerConfiguration.parallel.parallelExecutionConfiguration.parallelism)
    }

    @Override
    void visitSpec(SpecInfo spec) {
        // work-around for https://github.com/junit-team/junit5/issues/3108
        spec.allFeatures*.addIterationInterceptor {
            threadPool.submit(it::proceed).get()
        }
    }

    @Override
    void stop() {
        threadPool?.shutdown()
    }
}
```

**Comment by mufasa1** [comment]
> ```groovy
> IGlobalExten

> ```groovy
> class FixJunit3108Extension implements IGlobalExtension {
>     private static ExecutorService threadPool
> 
>     private final RunnerConfiguration runnerConfiguration
> 
>     FixJunit3108Extension(RunnerConfiguration runnerConfiguration) {
>         this.runnerConfiguration = runnerConfiguration
>     }
> 
>     @Override
>     void start() {
>         threadPool = Executors.newFixedThreadPool(runnerConfiguration.parallel.parallelExecutionConfiguration.parallelism)
>     }
> 
>     @Override
>     void visitSpec(SpecInfo spec) {
>         // work-around for https://github.com/junit-team/junit5/issues/3108
>         spec.allFeatures*.addIterationInterceptor {
>             threadPool.submit(it::proceed).get()
>         }
>     }
> 
>     @Override
>     void stop() {
>         threadPool?.shutdown()
>     }
> }
> ```

Hi, could you tell me where I can get this interface 'IGlobalExtension'? Is it in junit 5 anther issue branch?

**Comment by Vampire** [comment]
This is for Spock. I described above how you probably can do the same for Jupiter.

**Comment by igogoka** [comment]
> ```groovy
> class FixJunit3108Extension implements IGlobalExtension {
>     private static ExecutorService threadPool
> 
>     private final RunnerConfiguration runnerConfiguration
> 
>     FixJunit3108Extension(RunnerConfiguration runnerConfiguration) {
>         this.runnerConfiguration = runnerConfiguration
>     }
> 
>     @Override
>     void start() {
>         threadPool = Executors.newFixedThreadPool(runnerConfiguration.parallel.parallelExecutionConfiguration.parallelism)
>     }
> 
>     @Override
>     void visitSpec(SpecInfo spec) {
>         // work-around for https://github.com/junit-team/junit5/issues/3108
>         spec.allFeatures*.addIterationInterceptor {
>             threadPool.submit(it::proceed).get()
>         }
>     }
> 
>     @Override
>     void stop() {
>         threadPool?.shutdown()
>     }
> }
> ```

Thanks. I use your solution, but in Ui tests with Geb and Spock I have problem, that every feature (test in Spec)create new thread. How I can rewrite this method to create a new thread only for Spec ?
```groovy
 @Override
    void visitSpec(SpecInfo spec) {
        spec.allFeatures*.addIterationInterceptor {
            threadPool.submit(it::proceed).get()
        }

    } 
```

**Comment by Vampire** [comment]
It does not create a new thread for every features.
It uses a thread pool with a fixed size of which the threads are reused.

**Comment by igogoka** [comment]
![image](https://github.com/junit-team/junit5/assets/15260755/78502368-3159-4dfc-af1a-f769a2ad982e)
How I can set defaultExecutionMode = SAME_THREAD, beause it doesn't work from SpockConfig

**Comment by Vampire** [comment]
You are even more deviating from the topic here.
This is not a Spock-support channel and for every message you write, every watcher gets an e-mail. ;-)

**Comment by mufasa1** [comment]
@Vampire 

> This is for Spock. I described above how you probably can do the same for Jupiter.

Hi, I tried the description your wrote, but it didn't work. I would greatly appreciate it if you had a validated solution..

**Comment by Vampire** [comment]
I provided a full-blown example above, what more do you need?

**Comment by mufasa1** [comment]
> I provided a full-blown example above, what more do you need?

I mean for Junit, not for Spock. I tried your description like this
'@Override
public void interceptTestMethod(Invocation<Void> invocation,
        ReflectiveInvocationContext<Method> invocationContext,
        ExtensionContext extensionContext) throws Throwable {
    threadPool.submit(invocation::proceed).get();
}'

but it didn't work. What I need is a solution that has been validated.

**Comment by Vampire** [comment]
Works perfectly fine for me, except that in Jupiter `proceed` throws an exception, so you have to handle it.
But once you made the code compilable and ensured it is used, it does exactly what is intended.

**Comment by rturner-edjuster** [comment]
A couple of additional notes and a comment:
- The vintage execution also has the same problem
- The vintage execution seems to work for my tests and gets the expected result if you replace the `Executors.newWorkStealingPool` call with a call to `Executors.newFixedThreadPool` (albeit the related JUnit unit tests hang -- I wasn't able to quickly understand how they work to figure out why - so I disabled them locally to generate a snapshot build for testing with)

Given that the solution to this issue not imminent, I think it would be very useful to update the documentation around both the regular executor and the vintage executor to more clearly indicate that the current "limits" don't actually do what the documentation implies they do (or maybe it's just not clear to me, but I spent hours messing around with the settings, only to eventually start messing with the code to find that the choice of thread pooling isn't an intuitive one for executing tests in).

**Comment by OPeyrusse** [comment]
Thanks @rturner-edjuster to recenter this thread on the original topic.

**Comment by marcphilipp** [comment]
I've just submitted #5027. 

I've tried out the provided [sample project](https://github.com/OPeyrusse/junitforkjoinpool) you've provided against a locally built snapshot with the following changes:

```patch
Index: src/test/resources/junit-platform.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>ISO-8859-1
===================================================================
diff --git a/src/test/resources/junit-platform.properties b/src/test/resources/junit-platform.properties
--- a/src/test/resources/junit-platform.properties	(revision 070bce89a0e68bb87e31fb3f4f08f2b49b67c068)
+++ b/src/test/resources/junit-platform.properties	(date 1759738039643)
@@ -3,3 +3,4 @@
 junit.jupiter.execution.parallel.mode.classes.default = concurrent
 junit.jupiter.execution.parallel.config.strategy = fixed
 junit.jupiter.execution.parallel.config.fixed.parallelism=2
+junit.jupiter.execution.parallel.config.interceptor.class=org.junit.platform.engine.support.hierarchical.ParallelExecutionInterceptor$FixedThreadPoolForTests
Index: pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pom.xml b/pom.xml
--- a/pom.xml	(revision 070bce89a0e68bb87e31fb3f4f08f2b49b67c068)
+++ b/pom.xml	(date 1759731851552)
@@ -19,7 +19,7 @@
       <dependency>
         <groupId>org.junit</groupId>
         <artifactId>junit-bom</artifactId>
-        <version>5.8.2</version>
+        <version>6.1.0-SNAPSHOT</version>
         <type>pom</type>
         <scope>import</scope>
       </dependency>
```

As expected, now only two tests are started and only two threads from JUnit's `ForkJoinPool` are used.

@OPeyrusse Could you please run your actual production project against a snapshot to verify?

**Comment by OPeyrusse** [comment]
Let me find time to do it. Though if you said that you used my project and see it work, I expect to find the same result.
And to be sure, when you say "test with a snapshot", you mean that I must build junit locally and test it with my project?
This being my first bug report, I don't know what are the standard things to do

**Comment by marcphilipp** [comment]
> And to be sure, when you say "test with a snapshot", you mean that I must build junit locally and test it with my project?

Exactly! You should be able to run `./gradlew publishToMavenLocal` on the PR branch and then reference it as `6.1.0-SNAPSHOT` in your POM. We only publish snapshots from `main`, not from PR branches.

**Comment by OPeyrusse** [comment]
Yes, not pushing snapshots is very common. Let me try then on my side

**Comment by marcphilipp** [comment]
In the meantime, I've decided to take a different approach. As stated in https://github.com/junit-team/junit-framework/pull/5027#discussion_r2414406197, adding another thread pool solely for the purpose of hiding it from the `ForkJoinPool` seemed like a workaround rather than a solution. Therefore, I've spent the last couple of days re-implementing parallel execution with a regular thread pool while still supporting work stealing and resource locking. Moreover, the new implementation guarantees that no more than `parallelism` tasks will ever be executed in parallel.

Here it is:

- #5060

@OPeyrusse Please give the new PR a try when you have a chance.

**Comment by OPeyrusse** [comment]
Hello @marcphilipp , on my side, after building your latest version, the tests are still failing.
But it may be possible that changes are required.

I see that my code is naively testing with a static counter that no more than 20 tests are running in parallel. 
https://github.com/OPeyrusse/junitforkjoinpool/blob/main/src/main/java/org/example/Calculator.java#L26-L33

From this config, I expect to only have 2 tests running in parallel
https://github.com/OPeyrusse/junitforkjoinpool/blob/main/src/test/resources/junit-platform.properties#L4-L5

At the moment, with jdk 25, I see 
```console
[INFO] Running org.example.CalculatorTests
[INFO] [stdout] ForkJoinPool-1-worker-2 starting a new test. Running: 1
[INFO] [stdout] ForkJoinPool-1-worker-1 starting a new test. Running: 2
[INFO] [stdout] ForkJoinPool-1-worker-1 starting a new test. Running: 3
[INFO] [stdout] ForkJoinPool-1-worker-4 starting a new test. Running: 4
[INFO] [stdout] ForkJoinPool-1-worker-3 starting a new test. Running: 5
[INFO] [stdout] ForkJoinPool-1-worker-5 starting a new test. Running: 6
[INFO] [stdout] ForkJoinPool-1-worker-6 starting a new test. Running: 7
[INFO] [stdout] ForkJoinPool-1-worker-7 starting a new test. Running: 8
[INFO] [stdout] ForkJoinPool-1-worker-8 starting a new test. Running: 9
[INFO] [stdout] ForkJoinPool-1-worker-9 starting a new test. Running: 10
[INFO] [stdout] ForkJoinPool-1-worker-10 starting a new test. Running: 11
[INFO] [stdout] ForkJoinPool-1-worker-12 starting a new test. Running: 12
[INFO] [stdout] ForkJoinPool-1-worker-11 starting a new test. Running: 13
[INFO] [stdout] ForkJoinPool-1-worker-13 starting a new test. Running: 14
[INFO] [stdout] ForkJoinPool-1-worker-14 starting a new test. Running: 15
[INFO] [stdout] ForkJoinPool-1-worker-15 starting a new test. Running: 16
[INFO] [stdout] ForkJoinPool-1-worker-17 starting a new test. Running: 17
[INFO] [stdout] ForkJoinPool-1-worker-16 starting a new test. Running: 18
[INFO] [stdout] ForkJoinPool-1-worker-19 starting a new test. Running: 19
[INFO] [stdout] ForkJoinPool-1-worker-18 starting a new test. Running: 20
[INFO] [stdout] ForkJoinPool-1-worker-20 starting a new test. Running: 21
[INFO] [stdout] Too many tests running in parallel: 21
[INFO] [stdout] ForkJoinPool-1-worker-21 starting a new test. Running: 22
[INFO] [stdout] Too many tests running in parallel: 22
```
However, unlike the logs at the start of the ticket, I can see that no worker starts again and again new tests. It just seems that the engine does not comply with the limit of two threads max

**Comment by mpkorstanje** [comment]
> ForkJoinPool-1-worker-2 starting

@OPeyrusse you'll also want to add `junit.jupiter.execution.parallel.executor=org.junit.platform.engine.support.hierarchical.ConcurrentHierarchicalTestExecutorService` otherwise Jupiter will use the FokJoinPool executor instead of the threadpool based version.

**Comment by OPeyrusse** [comment]
It does not seem to work for me. At the moment, I have the following, based on your comment @mpkorstanje 
```ini
junit.jupiter.execution.parallel.enabled = true
junit.jupiter.execution.parallel.mode.default = concurrent
junit.jupiter.execution.parallel.mode.classes.default = concurrent
junit.jupiter.execution.parallel.config.strategy = fixed
junit.jupiter.execution.parallel.config.fixed.parallelism=2
junit.jupiter.execution.parallel.executor=org.junit.platform.engine.support.hierarchical.ConcurrentHierarchicalTestExecutorService
```
And I am still seeing the same error.
Does it work on your end? If yes, I can start debugging it and enable more logs, if you know some of them

**Comment by mpkorstanje** [comment]
Yes. I'm guessing you are working with a snapshot from the wrong PR. 

The expected output looks like something like this:

```
junit-1-worker-1 starting a new test. 
```

**Comment by OPeyrusse** [comment]
Indeed, I went to the PR mentioning my message but did not go to the second PR, for the other message. Testing it 🥁

**Comment by OPeyrusse** [comment]
And it is working 🚀
